### ISSUE TITLE
Is there a build-in way to map rare labels to the same index?

### ISSUE THREAD

**Issue by XinliYu** [allenai/allennlp#3053]
For example, NER, I have more than 10000 labels, some only occurs two or three times. Is there a build-in way to map these rare labels to the same index? Just like tokens, the `max_vocab_size` excludes rare tokens from the vocabulary so that they are mapped to the "@@UNKNOWN@@", but it seems not working for the labels. Thanks!

**Comment by DeNeutoy** [comment]
Unfortunately not - you should just do this in your dataset reader before creating the `Field`.

**Comment by matt-gardner** [comment]
@XinliYu, I think what you want is to just use our vocabulary configuration for your labels.  You need your label namespace to include an OOV token (easiest way to accomplish this is to have it not include the word "labels" or "tags" in the namespace, then we add OOV token by default), and you need to set the min count for that namespace in your vocabulary settings.  Then any label that gets seen fewer than that number of times gets mapped to the OOV token.

**Comment by schmmd** [comment]
Closing as this is answered--but please follow up if this doesn't answer your problem, or open a new issue if you have a different question.

**Comment by XinliYu** [comment]
> @XinliYu, I think what you want is to just use our vocabulary configuration for your labels. You need your label namespace to include an OOV token (easiest way to accomplish this is to have it not include the word "labels" or "tags" in the namespace, then we add OOV token by default), and you need to set the min count for that namespace in your vocabulary settings. Then any label that gets seen fewer than that number of times gets mapped to the OOV token.

Thanks @matt-gardner! Is there a tutorial about how to manipulate the vocabulary, or any reference would be great!

**Comment by matt-gardner** [comment]
We used to have a notebook tutorial on this, but it's not rendering for me right now: https://github.com/allenai/allennlp/blob/v0.7.0/tutorials/notebooks/vocabulary.ipynb.  Looking at the docs for `Vocabulary` is probably your best bet at this point.  We're hoping to have some better intro material in the next couple of months.

**Comment by XinliYu** [comment]
Thanks, a quick question is how to set the oov for the labels. I build the vocabulary using `Vocabulary.from_instances`, then it builds a vocabulary for `labels`, but the OOV token  `@@UNKNOWN@@` is not in it. An then it throws an error during indexing when it encounters the oov label.

The `from_instances` method does not seem to have an option to specify the oov token.

**Comment by XinliYu** [comment]
I did set an min count on the `labels` namespace, and it is working, the number of label to index mappings are reduced in the dictionary. But no oov is in there.

**Comment by matt-gardner** [comment]
tl;dr: easiest way is to set your namespace to something that doesn't end in "labels" or "tags".

To find out why that's the answer, read [here](https://github.com/allenai/allennlp/blob/5c64f9d01ef39e3398372ebe4f19f864691679c0/allennlp/data/vocabulary.py#L29-L63) and [here](https://github.com/allenai/allennlp/blob/5c64f9d01ef39e3398372ebe4f19f864691679c0/allennlp/data/vocabulary.py#L170-L183).  Reading the second one will show you another thing you can configure to make this work for `labels`, instead of changing your namespace.
